mdata/cache/accnt/flat_accnt.go

```diff
@@ -0,0 +1,310 @@
+		usg = 100 * a.total / a.maxSize
+	}
+
+	log.Info("--- Chunk Cache stats for the past %dns ---", duration)
+	log.Info("current size/max size %d/%d (%d%%)", a.total, a.maxSize, usg)
+	log.Info("metric complete %d", a.stats.complt_met)
+	log.Info("metric misses   %d", a.stats.miss_met)
+	log.Info("metric partials %d", a.stats.part_met)
+	log.Info("metric adds     %d", a.stats.add_met)
+	log.Info("metric evicts   %d", a.stats.evict_met)
+	log.Info("chunk hits      %d", a.stats.hit_chnk)
+	log.Info("chunk adds      %d", a.stats.add_chnk)
+	log.Info("chunk evicts    %d", a.stats.evict_chnk)
+
+	a.stats.complt_met = 0
+	a.stats.miss_met = 0
+	a.stats.part_met = 0
+	a.stats.add_met = 0
+	a.stats.evict_met = 0
+	a.stats.hit_chnk = 0
+	a.stats.add_chnk = 0
+	a.stats.evict_chnk = 0
+}
+
+func (a *FlatAccnt) eventLoop() {
+	for {
+		select {
+		case <-a.statsTicker.C:
+			a.statPrintReset()
+		case event := <-a.eventQ:
+			switch event.t {
+			case evnt_add_chnk:
+				payload := event.pl.(*AddPayload)
+				a.add(payload.metric, payload.ts, payload.size)
+				a.stats.add_chnk++
+				a.lru.touch(
+					EvictTarget{
+						Metric: payload.metric,
+						Ts:     payload.ts,
+					},
+				)
+			case evnt_hit_chnk:
+				payload := event.pl.(*HitPayload)
+				a.stats.hit_chnk++
+				a.lru.touch(
+					EvictTarget{
+						Metric: payload.metric,
+						Ts:     payload.ts,
+					},
+				)
+			case evnt_miss_met:
+				a.stats.miss_met++
+			case evnt_part_met:
+				a.stats.part_met++
+			case evnt_complt_met:
+				a.stats.complt_met++
+			}
+
+			// evict until we're below the max
+			for a.total > a.maxSize {
+				a.evict()
+			}
+		}
+	}
+}
+
+func (a *FlatAccnt) add(metric string, ts uint32, size uint64) {
+	var met *FlatAccntMet
+	var ok bool
+
+	if met, ok = a.metrics[metric]; !ok {
+		met = &FlatAccntMet{
+			total:  0,
+			chunks: make(map[uint32]uint64),
+		}
+		a.metrics[metric] = met
+		a.stats.add_met++
+	}
+
+	if _, ok = met.chunks[ts]; ok {
+		// we already have that chunk
+		return
+	}
+
+	met.chunks[ts] = size
+	met.total = met.total + size
+	a.total = a.total + size
+}
+
+func (a *FlatAccnt) evict() {
+	var met *FlatAccntMet
+	var targets []uint32
+	var ts uint32
+	var size uint64
+	var ok bool
+	var e interface{}
+	var target EvictTarget
+
+	e = a.lru.pop()
+
+	// got nothing to evict
+	if e == nil {
+		return
+	}
+
+	// convert to EvictTarget otherwise
+	target = e.(EvictTarget)
+
+	if met, ok = a.metrics[target.Metric]; !ok {
+		return
+	}
+
+	for ts = range met.chunks {
+		// if we have chronologically older chunks we add them
+		// to the evict targets to avoid fragmentation
+		if ts <= target.Ts {
+			targets = append(targets, ts)
+		}
+	}
+
+	sort.Sort(Uint32Asc(targets))
+
+	for _, ts = range targets {
+		size = met.chunks[ts]
+		met.total = met.total - size
+		a.total = a.total - size
+		a.stats.evict_chnk++
+		a.evictQ <- &EvictTarget{
+			Metric: target.Metric,
+			Ts:     ts,
+		}
+		delete(met.chunks, ts)
+	}
+
+	if met.total <= 0 {
```

Comment:
yes, this might as well be `== 0`, anything that's `< 0` would be one of those "impossible situations".